### ISSUE TITLE
Consider: Which CI tests should run with -fno-exceptions

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp#2525]
We [decided](https://github.com/googleapis/google-cloud-cpp/blob/master/doc/adr/2019-01-04-error-reporting-with-statusor.md) that our libraries will not require exceptions to work. That is, a user should be able to compile with `-fno-exceptions` and still be able to use our library just fine. All of our error reporting is done via `Status` or `StatusOr<T>` objects instead of throwing exceptions. This decision does not *prohibit* our users from using exceptions themselves, it simply gives them the choice about whether or not they use exceptions.

So since our library must work whether it's compiled with `-fexceptions` (the default) or with `-fno-exceptions` we should think about what flags our tests should use. The options I see are:

1. (Status Quo) **Compile and run most tests with `-fexceptions`, and one (or a small few) with `-fno-exceptions`**.  This is the default setting for compilers, it is the setting used by most C++ users ([50+%](https://isocpp.org/blog/2018/03/results-summary-cpp-foundation-developer-survey-lite-2018-02) of users compile with exceptions enabled), and it is the setting required for proper/standard C++ (i.e., C++ is only defined **with** exceptions enabled; disabling exceptions is a compiler-supported non-standard dialect of C++).

2. **Compile and run most tests with `-fno-exceptions` and one (or a small few) with `-fexceptions`**. This puts most of our testing in the no-exceptions world, which we claim our library supports. One could argue that if the library works with `-fno-exceptions`, then it will work with exceptions because that's just adding features (throw, try, catch, etc) that we're not using. To a rough approximation, that reasoning makes some sense, but there are cases that it doesn't cover.

3. **Run *all tests* both with and without exceptions**. This is the most thorough testing and likely the ideal situation. However, it will close to double our testing requirements. However, if we can "afford" to run all these tests, it may be worth it.

**Comment by devjgm** [comment]
I'm somewhat partial to option 3, if we can make it work w/ resources (including the time it takes to run the tests on a PR).

**Comment by remyabel** [comment]
By "tests" are you referring to the examples? From my understanding, the noex build already will compile the unit/integration tests with `-fno-exceptions` but benchmarks/examples are compiled only if exceptions are enabled.

**Comment by devjgm** [comment]
For some context, here are the builds we currently have:

# Travis:
1. Coverage build
2.  Checkstyle and generate docs (push to gh-pages?)

# Kokoro:
3. MacOS
4. Windows
5. Ubuntu
6. CentOS7
7. Bazel Linux
8. Asan
9. UBSAN
10. Shared
11. NoEx
12. LibCXX
13. ClangTidy
14. Check ABI
* Install tests:
15. openSUSE: Tumbleweed
16. openSUSE: Leap
17. Ubuntu: Trusty
18. Ubuntu: Xenial
19. Ubuntu: Bionic
20. Fedora: 29
21. Debian: 9
22. Centos: 7

Now let me assume that the different build configurations we'd like are something like:
* gcc
* clang
* `-fexceptions`
* `-fno-exceptions`
* bazel
* cmake
* (maybe also shared vs static?)

Perhaps we could run all 6 of those configurations on {MacOS, Ubuntu, CentOS7}, resulting in 3 * 6 = 18 new build configurations. But we'd also be able to remove at least 7 existing builds, possibly more if we decide to run all of some flavor of build with ubsan/asan enabled.

In short, doing option (3) above would not *double* our required tests. It would increase it, but perhaps by about 11-ish tests

**Comment by devjgm** [comment]
@remyabel no, I'm referring to build configurations, not just the examples.

**Comment by devjgm** [comment]
We could even consider taking bazel out of the test matrix and so that it doesn't need to run on every platform in every configuration. It could just be run as a one-off build, like it is now. I think that'd be OK.

**Comment by coryan** [comment]
You may want to check go/cbt-cpp:build-matrix and go/gcs-cpp:build-matrix-trix.

I argue that we cannot test all possible combinations of `OS`, `DISTRO`, `BUILD-SYSTEM`, `COMPILER`, `COMPILER-VERSION`, `COMPILER-FLAGS` (there is more than `-fno-exceptions`, things like `-O3` vs. `-g`, `-fno-rtti`), `LINK MODE`, `INSTALLED-vs-SOURCE`.  The numbers get large, even for Google, real fast like :grimacing: 

The approach I propose in those docs is to get coverage by making sure that we have at least one build with shared libraries, and at least one build with `-fno-exceptions`, and at least one build with each version of GCC, and so forth.

That is not to say that we could improve the build matrix, and we probably have gaps in the coverage.

**Comment by remyabel** [comment]
> @remyabel no, I'm referring to build configurations, not just the examples.

OK, my bad. I misunderstood what you meant by "test".

We can eliminate "Ubuntu: Trusty" as that will stop receiving LTS support April 30th. 

If Kokoro has support for a multi-dimensional build matrix, that would probably simplify things.

**Comment by devjgm** [comment]
@coryan I agree that we can't test all possible combinations. You're right that there are simple too many flags and combinations to do this. Indeed, my math above was wrong. It's not 6 configurations I was proposing, it was 2 * 2 * 2 = 8 combinations, times 3 OSes, equals 24 (not 18). So I agree it's more than I originally thought.

But I still think we may be able to do a little more to expand our no-exceptions testing without compromising our exceptions testing. Some observations:

* We don't need to test with all compiler flag combinations. We pick a set of flags that we like and support (perhaps a fairly strict set, and [perhaps similar to Abseil's](https://abseil.io/docs/cpp/platforms/compilerflags)) and simply use those.
* The `-fno-exceptions`/`-fexceptions` flags are different though. These put all the code into a different "mode" if you will; one mode where we're using a non-standard dialect of C++ (noex), that we claim to support. I think it's important that we test fairly thoroughly in this mode.
* I don't think we need to test with each version of gcc. I didn't see that proposed in the doc you referenced, and I don't see us doing that today (but perhaps I'm not looking in the right place).
* I also think that testing with clang vs gcc is pretty important. This not only changes the compiler, but it also changes the standard library (by default; libstdc++ vs libc++).

Overall, I think we can expand our testing with `-fno-exceptions` today by by only needing to add a handful of builds. If this is still too much for us to test, we might consider if more testing with `-fno-exceptions` is more important than some other tests we're currently doing. Like maybe testing on some Linux distro is less important and can be dropped[1]? Just thinking/discussing.

[1]: What linux distros are popular enough for us to test on them? Are there actual distro usage numbers somewhere? [Google trends says this](https://trends.google.com/trends/explore?geo=US&q=debian,ubuntu,centos,opensuse,fedora), which is just based on search volume, so not exactly what we're looking for.

**Comment by coryan** [comment]
> But I still think we may be able to do a little more to expand our no-exceptions testing without compromising our exceptions testing.

I think we agree on that.

> It's not 6 configurations I was proposing, it was 2 * 2 * 2 = 8 combinations, times 3 OSes, equals 24 (not 18). So I agree it's more than I originally thought.

We currently have about 22 builds, so we are talking about doubling the number of builds. So while I still agree we need more builds with `-fno-exceptions`, maybe there is room to optimize this list?

For example: I think we only test with GCC right now, it is probably a good idea to test with Clang, and MSVC, and the macOS version of Clang.  That is 4 more builds, and it gives us the coverage we need maybe?

Again, I fully agree that we need more builds with `-fno-exceptions`.  I am opposed to making `-fno-exceptions` the default or the majority of our builds. Well, if we had infinite bandwidth I would probably say, sure, build with `-fexceptions` and `-fno-exceptions` everywhere.  But given that we have (somewhat) limited bandwidth, the default build on a OS+compiler combo should be "with the default flags", because that is (the most likely) experience for our users on that OS+compiler combo.

----

Some slight off-topic things below:

> I also think that testing with clang vs gcc is pretty important. This not only changes the compiler, but it also changes the standard library (by default; libstdc++ vs libc++).

Actually it does not, on most distros you still get `libstdc++` when you compile with Clang.  You need to do special things to get `libc++`.  We do have one build that does those special things, thanks to @remyabel:

https://github.com/googleapis/google-cloud-cpp/blob/cd3bc6b4eb20cf5c22c5e981ad0314969493a61a/ci/kokoro/docker/libcxx.cfg#L21

> I don't think we need to test with each version of gcc. I didn't see that proposed in the doc you referenced, and I don't see us doing that today (but perhaps I'm not looking in the right place).

We do not test every version. But we compile with different distros and distro versions, which have different versions of GCC anyway.

> The -fno-exceptions/-fexceptions flags are different though. These put all the code into a different "mode" if you will;

So does `-fno-rtti`.  And `-DNDEBUG` changes what `assert()` does, and we have dependencies that do nice things like this (note this is a *header*, but it *probably* does not affect us):

https://github.com/grpc/grpc/blob/b41a44dc2e001358602e8e885c674ddc49387edd/src/core/lib/iomgr/closure.h#L94

> [1]: What linux distros are popular enough for us to test on them? Are there actual distro usage numbers somewhere? Google trends says this, which is just based on search volume, so not exactly what we're looking for.

I think the interesting number is what Linux distros are popular on GCE, I remember seeing a presentation about it, cannot find it anymore :disappointed:

**Comment by devjgm** [comment]
> > But I still think we may be able to do a little more to expand our no-exceptions testing without compromising our exceptions testing.
> 
> I think we agree on that.
> 
> > It's not 6 configurations I was proposing, it was 2 * 2 * 2 = 8 combinations, times 3 OSes, equals 24 (not 18). So I agree it's more than I originally thought.
> 
> We currently have about 22 builds, so we are talking about doubling the number of builds. So while I still agree we need more builds with `-fno-exceptions`, maybe there is room to optimize this list?

It would be less than doubling the number of builds, because these new builds would replace some of the existing builds. Additionally, we would not need to increase any of our 8 install/test-readme builds.

> 
> For example: I think we only test with GCC right now, it is probably a good idea to test with Clang, and MSVC, and the macOS version of Clang. That is 4 more builds, and it gives us the coverage we need maybe?
> 
> Again, I fully agree that we need more builds with `-fno-exceptions`. I am opposed to making `-fno-exceptions` the default or the majority of our builds.

Even though I originally proposed this idea in a meeting yesterday, I think I now agree with you that noex should not be our default. So I want to be clear that I'm not proposing that, and specifically, I am not in favor of option (2) from my OP.

> Well, if we had infinite bandwidth I would probably say, sure, build with `-fexceptions` and `-fno-exceptions` everywhere. But given that we have (somewhat) limited bandwidth, the default build on a OS+compiler combo should be "with the default flags", because that is (the most likely) experience for our users on that OS+compiler combo.
> 
> Some slight off-topic things below:
> 
> > I also think that testing with clang vs gcc is pretty important. This not only changes the compiler, but it also changes the standard library (by default; libstdc++ vs libc++).
> 
> Actually it does not, on most distros you still get `libstdc++` when you compile with Clang. You need to do special things to get `libc++`. We do have one build that does those special things, thanks to @remyabel:
> 
> [google-cloud-cpp/ci/kokoro/docker/libcxx.cfg](https://github.com/googleapis/google-cloud-cpp/blob/cd3bc6b4eb20cf5c22c5e981ad0314969493a61a/ci/kokoro/docker/libcxx.cfg#L21)
> 
> Line 21 in [cd3bc6b](/googleapis/google-cloud-cpp/commit/cd3bc6b4eb20cf5c22c5e981ad0314969493a61a)
> 
>  value: "libcxx" 

Ahh, ok, thanks.

> > I don't think we need to test with each version of gcc. I didn't see that proposed in the doc you referenced, and I don't see us doing that today (but perhaps I'm not looking in the right place).
> 
> We do not test every version. But we compile with different distros and distro versions, which have different versions of GCC anyway.
> 
> > The -fno-exceptions/-fexceptions flags are different though. These put all the code into a different "mode" if you will;
> 
> So does `-fno-rtti`. And `-DNDEBUG` changes what `assert()` does, and we have dependencies that do nice things like this (note this is a _header_, but it _probably_ does not affect us):

I think these are quite different. We make a specific claim about using our library with and without exceptions. We make no such claim about `no-rtti`. I mean, of course we should work, but that's more like an implicit assumption. Our position on exceptions is different to the point where it affects how we design APIs.

Of course, given infinite bandwidth, we could test everything.... but in actuality, we have to draw the line somewhere. In my opinion, testing with and without exceptions falls on the side of "we should test it thoroughly both ways" because we make specific claims about this particular feature. Other compiler flags and defines fall on the "do the best you can" side.

> 
> https://github.com/grpc/grpc/blob/b41a44dc2e001358602e8e885c674ddc49387edd/src/core/lib/iomgr/closure.h#L94
> 
> > [1]: What linux distros are popular enough for us to test on them? Are there actual distro usage numbers somewhere? Google trends says this, which is just based on search volume, so not exactly what we're looking for.
> 
> I think the interesting number is what Linux distros are popular on GCE, I remember seeing a presentation about it, cannot find it anymore

Just a thought: Given that GCP customers can run whatever distros they want in their own containers on GCE and GKE and Cloud Run, etc, I'm not sure how much we should prioritize the distros-on-GCE numbers.... but absent any numbers (and I have none), I guess it doesn't matter much anyway :-)

**Comment by coryan** [comment]
> > We currently have about 22 builds, so we are talking about doubling the number of builds. So while I still agree we need more builds with `-fno-exceptions`, maybe there is room to optimize this list?
> 
> It would be less than doubling the number of builds, because these new builds would replace some of the existing builds. Additionally, we would not need to increase any of our 8 install/test-readme builds.

Ack.

> > For example: I think we only test with GCC right now, it is probably a good idea to test with Clang, and MSVC, and the macOS version of Clang. That is 4 more builds, and it gives us the coverage we need maybe?
> > Again, I fully agree that we need more builds with `-fno-exceptions`. I am opposed to making `-fno-exceptions` the default or the majority of our builds.
> 
> Even though I originally proposed this idea in a meeting yesterday, I think I now agree with you that noex should not be our default. So I want to be clear that I'm not proposing that, and specifically, I am not in favor of option (2) from my OP.

Ack.

I think it is time to turn this into a PR or something.  We both agree that we do need more builds with `-fno-exceptions`, I think we both agree that doubling is probably too many, we both agree that we could replace some builds.  Now it is a matter of actually writing the changes to the Kokoro configs?

> > So does `-fno-rtti`. And `-DNDEBUG` changes what `assert()` does, and we have dependencies that do nice things like this (note this is a _header_, but it _probably_ does not affect us):
> 
> I think these are quite different. We make a specific claim about using our library with and without exceptions. We make no such claim about `no-rtti`. I mean, of course we should work, but that's more like an implicit assumption. Our position on exceptions is different to the point where it affects how we design APIs.

Ack.

> 
> Of course, given infinite bandwidth, we could test everything.... but in actuality, we have to draw the line somewhere. In my opinion, testing with and without exceptions falls on the side of "we should test it thoroughly both ways" because we make specific claims about this particular feature. Other compiler flags and defines fall on the "do the best you can" side.

Sounds good.

**Comment by devjgm** [comment]
This issue is old and huge. I did not re-read the whole thing. But it's about CI test running with -fno-exceptions, and this seems like an issue that has long since been solved and is currently working correctly. So I suspect this issue is now either fixed or obsolete. Closing to clean things up.
